pkg/storegateway/bucket.go

```diff
@@ -727,54 +730,61 @@ func blockSeries(
 	return newBucketSeriesSet(res), reqStats, nil
 }
 
+func extractLabelsFromSeriesEntries(entries []seriesEntry) (result []labels.Labels) {
+	result = make([]labels.Labels, len(entries))
+	for i := range result {
+		result[i] = entries[i].lset
+	}
+	return
+}
+
 type seriesCacheEntry struct {
 	LabelSets   []labels.Labels
 	MatchersKey indexcache.LabelMatchersKey
 	Shard       sharding.ShardSelector
 }
 
-func fetchCachedSeries(ctx context.Context, userID string, indexCache indexcache.IndexCache, blockID ulid.ULID, matchers []*labels.Matcher, shard *sharding.ShardSelector, logger log.Logger) ([]seriesEntry, bool) {
+func fetchCachedSeries(ctx context.Context, userID string, indexCache indexcache.IndexCache, blockID ulid.ULID, matchers []*labels.Matcher, shard *sharding.ShardSelector, postings indexcache.PostingsKey, logger log.Logger) (seriesChunkRefsSet, bool) {
```

Comment:
I would prefer to not change the caching logic in the non-streaming implementation. I would add new utility functions and new functions on the cache interface for the new caching based on postings.